### PR TITLE
New style hard sigmoid

### PR DESCRIPTION
**Pull request by hvy**
New style hard sigmoid to support double backprop.

### CODE REVIEW

**Comment by unnonouno** [comment]
jenkins test this please

File: chainer/functions/activation/hard_sigmoid.py [hunk 1]
```diff
@@ -19,30 +19,49 @@ def check_type_forward(self, in_types):
...
+class HardSigmoidGrad(function_node.FunctionNode):
+
+    """Hard-sigmoid gradient function."""
+
+    def __init__(self, x):
+        self.x = x
+
+    def check_type_forward(self, in_types):
+        type_check.expect(in_types.size() == 1)
+        type_check.expect(in_types[0].dtype.kind == 'f')
```
Comment (unnonouno) [inline chainer/functions/activation/hard_sigmoid.py]: Check `in_types[0].dtype == self.x.dtype`

Comment (hvy) [reply chainer/functions/activation/hard_sigmoid.py]: Added this type check.

**Comment by hvy** [comment]
Thank you for the review. I added another dtype check to the gradient class.

**Comment by hvy** [comment]
Rebased on top of master.

**Comment by unnonouno** [comment]
LGTM

**Pull request merged**